### ISSUE TITLE
Worlds: Ask To Join Session Mode

### ISSUE THREAD

**Issue by GearBell** [Frooxius/NeosPublic#1084]
Can we have a new Ask To Join session mode for a world?
Context: Dante (im gonna use him alot in this example) started a good trend of a session being mish mashed into being an "ask to join" setting by setting user limit to start at one, the set world as public or friends only - this way it advertises on the Open Worlds sessions as joinable and okay to visit, but Dante has complete control over who joins cause no one can spontaniously join as it will say "unable to join, user limit". Someone asks in messaging, and you set the world to private, raise the user limit by one, send them an invite, wait until they join, then re-set the session as public or friends only to continue to advertise the world in session joiner.
Ive personally been using this alot now, and it works great, except for one flaw - when someone leaves and i get caught up with stuff, i forget to lower the user limit again since the person left, and because of that (without fail) someone waltzes right in without asking, because the forgotten user limit allowed them in.

I ask that we have a real Ask to Join session mode, so that it will advertise that the world is open in the session joiner that visitors are welcome, but if they click to join they will get a prompt saying "Ask Host to Join".
This is different from a closed private world session mode. private worlds are seen as a "do not disturb" thing, whereas an Ask to Join is more open and inviting and no one grumbles that you are "hiding in a private world".
An Ask to Join session mode will allow the community to be a bit more inviting for sessions, while still maintaining who joins. (this is regardless of the future Friends Groups, as Ask to Join is a generalized world control)

[currently this would only allow Friends, as you must be friends to message the host, but perhaps a far future update would be Anyone clicks to join and will get a prompt saying "Join Request Sent" and then the session host gets a prompt saying "<Username> requests to join. Accept/Deny?" and accept lets them in and Deny pretty much says "Request Canceled at this time" (to sound less aggressive]
Thoughts?

**Comment by mralext20** [comment]
Sounds like a good way to reuse the existing discord invite system. (Older version seen here: https://this-is-a.stale-me.me/i/tp2y.jpg )

**Comment by GearBell** [comment]
nice! thats half the mechanism there, now just for a publically advertised closed session. XD

**Comment by BlazeVR** [comment]
I personally don't like this "Ask to join" as an additional session mode, it's kinda resembling the same orange status culture from VRC (but this time you can see their location) which people generally despise. 

However for your issue we have this update from some months ago:
> Host sending an invite to user will now allow the invited user to bypass the session Max User limit (suggested by Shifty | Quality Control Lead)

So in theory you could continue with just letting it on public/friends again where people can see your world and have to message, but without forgetting to lower/raise the user limit.

(Btw the Discord invite system is pretty handy, forgot it was a thing haha)

**Comment by Frooxius** [comment]
Hmm, this could be possible, but the main question is how exactly to go about it. It would be best if it was indicated directly in the UI that you have to request join - e.g. instead of "Join" button it would say "Request Join/Invite" and handle it for you. 

However should sessions like this show up for everyone as public? Or just friends? If someone is not a friend, they won't be able to send you a message, but the system could be made so they can send a request for invite regardless.

Also would it be important for them to include a message in the request to join?

There's one thing with the current system that helps though - you don't actually need to raise or lower the user limit when sending out invites. As host, if you send out an invite, that automatically makes that person bypass the session user limit. E.g. you could set your session to just 1 user and keep it public, then just send them an invite. They'll be able to join despite the limit.

**Comment by GearBell** [comment]
Yes the idea is to let the world be seen in the session joiner. if the sending an invite request thing to the host is indeed possible, then itll work regardless of friend or not? no message needed i feel, just the join request. if they are a friend and they have a specific reason for wanting to join then they can always message. but the join prompt alex showed in the image looks just about right!
the whole point is to let the host have the choice of "not hiding in a private world" but still have control over who joins. so simply go to session menu, modes (or whatever its called) and click session mode as Ask to Join!

**Comment by GearBell** [comment]
> I personally don't like this "Ask to join" as an additional session mode, it's kinda resembling the same orange status culture from VRC (but this time you can see their location) which people generally despise.
> 
> However for your issue we have this update from some months ago:
> 
> > Host sending an invite to user will now allow the invited user to bypass the session Max User limit (suggested by Shifty | Quality Control Lead)
> 
> So in theory you could continue with just letting it on public/friends again where people can see your world and have to message, but without forgetting to lower/raise the user limit.
> 
> (Btw the Discord invite system is pretty handy, forgot it was a thing haha)

unfortunatly that will work with friends only and theyd have to message you everytime, and the botification system isnt the best. (prompt is out of viewpoint and forget to check notifications). with a "user wants to join" popup it allows anyone, even nonfriends to send a request. This whole thing is to help neos open up a bit more, rather than everyone already hiding in private worlds.

**Comment by TehTurk** [comment]
I meant isn't this just no different from an Invite Only World? Except with an Ad Feature to notify host that someone wants to join and the main thing you want is the ability to reverse how it's shown in the session lists too is what it sounds like.  (That'd be somewhat neat). I'm all for just the distinction can be Confused once we have Invite/Private, RequestOnly, Friends, Friends+(At some Point), Registered Users, Anyone. Seems like it'd be easier what you want your session to have/do vs the distinction between them.

**Comment by StarfishHidari** [comment]
I often have to cap my worlds because my internet cannot handle more than 8 people, so I often cap it at 4 people and put in the title for people to ask me for an invite. It would be a lot easier if they could just request an invite possibly with a message that I can accept (thus sending them an invite to the world), or I can ignore, dismissing the request.
